[inductor] Refactor memory management code in wrapper codegen

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96768

Summary: use inheritance to simplify CppWrapperCodeGen and to prepare for AOT codegen

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

desertfire committed:
[inductor] Refactor memory management code in wrapper codegen

[ghstack-poisoned]

ngimel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -213,53 +142,13 @@ def codegen(self, code: IndentedBuffer):
-                del_func=lambda name: f"; {name}.reset()",
-                declare="auto ",
-                ending=";",
-                as_strided="at::as_strided",
```
do we no longer need to generate cpp `at::as_strided` calls?

desertfire replied:
We still need `at::as_strided`, but now it is generated in make_buffer_reuse which works for both WrapperCodeGen and CppWrapperCodeGen.

desertfire commented:
@pytorchbot merge -r

desertfire committed:
Update on "[inductor] Refactor memory management code in wrapper codegen"


Summary: use inheritance to simplify CppWrapperCodeGen and to prepare for AOT codegen

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge -r

desertfire committed:
Update on "[inductor] Refactor memory management code in wrapper codegen"


Summary: use inheritance to simplify CppWrapperCodeGen and to prepare for AOT codegen

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge

desertfire committed:
Update on "[inductor] Refactor memory management code in wrapper codegen"


Summary: use inheritance to simplify CppWrapperCodeGen and to prepare for AOT codegen

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
